### DISCUSSION TITLE
How to check the dirty state of a property from Javascript?

### DISCUSSION THREAD

**Discussion by livelogic** [livewire/livewire#6812] [1. Help] [answered]
Some properties of a Livewire component are accessible on the Alpine/Javascript side via `this.$wire`, but apparently this is not helpful to determine if a bound property is dirty.

Consider an Alpine component that allows binding via `wire:model` with any modifier, for an entangled property: when the modifiers include _blur_, we may listen to _blur_ events from the input element and trigger the property 'commit' via `this.$wire.set(....)`, but only if that property is dirty.

The current documentation lists some global Javascript hooks that can be used to expose, among other things, the canonical state of the component which may be compared against the current (Alpine) state, but this seem like a very large trip for the purpose.

**Comment by alexanderchr** [comment, marked as answer]
A dirty (lol) solution is to add the following to your component template

```html
<div id="dirty-monitor" wire:dirty.class="dirty"></div>
```

and then:

```js
const dirtyMonitor = document.getElementById('dirty-monitor');
const componentIsDirty = dirtyMonitor && dirtyMonitor.classList.contains('dirty'));
```

Combine with `wire:target` if you need to check dirty state of individual properties. Agree that dirty state should be accessible on `$wire`.

**Reply by livelogic** [reply]
Effective solution :)
And yes, i'd be a lot happier too with some APIs like `$wire.isDirty(propName)`.

**Comment by michael-lajos** [comment]
Another way to solve it is to compare the old value yourself. You can access the last known server state by using the canonical state.

```$wire.__instance.canonical.[YOUR PROPERTY NAME HERE]``` 

Then you can use that to compare it to the new value. For more info see: https://livewire.laravel.com/docs/javascript#the-component-object

**Reply by dllsystem** [reply]
```
const component = $wire.__instance
const isDirty = JSON.stringify(component.canonical) !== JSON.stringify(component.reactive)
```
